Object_Detection_Nanodet

Sidd1609 opened the pull request:
Google Summer of Code'22
Contributor: Sri Siddarth Chakaravarthy 
Project: Lightweight Object Detection
Model: Nanodet 

Submitting pull requests for Mid-term evaluations. 

COMPLETED:

- [x] Nanodet ONNX model (FP16)
- [x] ReadMe
- [x] License
- [x] Sample Images and Results 
- [x] Video Inference

TO DO:

- [ ] Model Quantization

fengyuentau reviewed (changes requested):
Thank you 👍 Please take a look at the comments below.

By the way, can you remove the axis of the result images for better visual effects?

Sidd1609 commented:
Ok yuantao will update these changes and update the pull request.

On Mon, Jul 25, 2022 at 12:48 PM Yuantao Feng ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> Thank you 👍 Please take a look at the comments below.
>
> By the way, can you remove the axis of the result images for better visual
> effects?
> ------------------------------
>
> In models/object_detection_nanodet/Nanodet_m_plus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928513752>:
>
> > @@ -0,0 +1,158 @@
>
> +import cv2
>
> +import numpy as np
>
> +import argparse
>
> +import time
>
> +import matplotlib.pyplot as plt
>
> +
>
> +class nanodet():
>
>
> I think it should be named as NanoDet. So class NanoDet:
> ------------------------------
>
> In models/object_detection_nanodet/Nanodet_m_plus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928514977>:
>
> > +
>
> +class nanodet():
>
> +    def __init__(self, prob_threshold=0.35, iou_threshold=0.6):
>
> +        with open('coco.names', 'rt') as f:
>
> +            self.classes = f.read().rstrip('\n').split('\n')
>
> +
>
> +        self.num_classes = len(self.classes)
>
> +        self.strides = (8, 16, 32, 64)
>
> +        self.image_shape = (416, 416)
>
> +        self.reg_max = 7
>
> +        self.prob_threshold = prob_threshold
>
> +        self.iou_threshold = iou_threshold
>
> +        self.project = np.arange(self.reg_max + 1)
>
> +        self.mean = np.array([103.53, 116.28, 123.675], dtype=np.float32).reshape(1, 1, 3)
>
> +        self.std = np.array([57.375, 57.12, 58.395], dtype=np.float32).reshape(1, 1, 3)
>
> +        self.net = cv2.dnn.readNet('nanodet-plus-m-1.5x_416.onnx')
>
>
> The model should be renamed to
> object_detection_nanodet-plus-m-1.5x-416.onnx
> ------------------------------
>
> In models/object_detection_nanodet/Nanodet_m_plus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928516320>:
>
> > +                img = cv2.copyMakeBorder(img, 0, 0, left, self.image_shape[1] - neww - left, cv2.BORDER_CONSTANT,
>
> +                                         value=0)  # add border
>
> +            else:
>
> +                newh, neww = int(self.image_shape[0] * hw_scale), self.image_shape[1]
>
> +                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
>
> +                top = int((self.image_shape[0] - newh) * 0.5)
>
> +                img = cv2.copyMakeBorder(img, top, self.image_shape[0] - newh - top, 0, 0, cv2.BORDER_CONSTANT, value=0)
>
> +        else:
>
> +            img = cv2.resize(srcimg, self.image_shape, interpolation=cv2.INTER_AREA)
>
> +        return img, newh, neww, top, left
>
> +
>
> +    def detect(self, srcimg):
>
> +        img, newh, neww, top, left = self.resize_image(srcimg)
>
> +        img = self._normalize(img)
>
> +        blob = cv2.dnn.blobFromImage(img)
>
> +        print(blob.shape)
>
>
> Do not print in the inference function.
> ------------------------------
>
> In models/object_detection_nanodet/Nanodet_m_plus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928518766>:
>
> > +        img = self._normalize(img)
>
> +        blob = cv2.dnn.blobFromImage(img)
>
>
> These two lines should be placed in a function called pre_process. Do
> something like this:
>
> def pre_process(self, img):
>
>     img = self._normalize(img)
>
>     blob = cv2.dnn.blobFromImage(img)
>
>     return blob
>
> ------------------------------
>
> In models/object_detection_nanodet/Nanodet_m_plus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928521450>:
>
> > +                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
>
> +                top = int((self.image_shape[0] - newh) * 0.5)
>
> +                img = cv2.copyMakeBorder(img, top, self.image_shape[0] - newh - top, 0, 0, cv2.BORDER_CONSTANT, value=0)
>
> +        else:
>
> +            img = cv2.resize(srcimg, self.image_shape, interpolation=cv2.INTER_AREA)
>
> +        return img, newh, neww, top, left
>
> +
>
> +    def detect(self, srcimg):
>
> +        img, newh, neww, top, left = self.resize_image(srcimg)
>
> +        img = self._normalize(img)
>
> +        blob = cv2.dnn.blobFromImage(img)
>
> +        print(blob.shape)
>
> +        self.net.setInput(blob)
>
> +
>
> +        outs = self.net.forward(self.net.getUnconnectedOutLayersNames())
>
> +        det_bboxes, det_conf, det_classid = self.post_process(outs)
>
>
> We need to put pre-process, forward, and post-process in a function called
> infer for benchmarking. Do something like this:
>
> def infer(self, img):
>
>     input_blob = self.pre_process(img)
>
>
>
>     self.net.setInput(input_blob)
>
>     outs = self.net.forward(self.net.getUnconnectedOutLayersNames())
>
>
>
>     bboxes, conf, classid = self.post_process(outs)
>
>     return bboxes, conf, classid
>
>
> def detect(self, img):
>
>     img, newh, neww, top, left = self.resize_image(img)
>
>
>
>     det_bboxes, det_conf, det_classid = self.infer(img)
>
>
>
>     return det_bboxes, det_conf, det_classid
>
> self.detect is kept because for now we do not want to have the time of
> resizing in benchmarking and it is used in demo.py to produce correct
> results.
> ------------------------------
>
> In models/object_detection_nanodet/Nanodet_m_plus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928522492>:
>
> > +            bboxes_mlvl /= scale_factor
>
> +        scores_mlvl = np.concatenate(scores_mlvl, axis=0)
>
> +
>
> +        bboxes_wh = bboxes_mlvl.copy()
>
> +        bboxes_wh[:, 2:4] = bboxes_wh[:, 2:4] - bboxes_wh[:, 0:2]  ####xywh
>
> +        classIds = np.argmax(scores_mlvl, axis=1)
>
> +        confidences = np.max(scores_mlvl, axis=1)  ####max_class_confidence
>
> +
>
> +        indices = cv2.dnn.NMSBoxes(bboxes_wh.tolist(), confidences.tolist(), self.prob_threshold, self.iou_threshold)
>
> +        if len(indices)>0:
>
> +            bboxes_mlvl = bboxes_mlvl[indices[:]]
>
> +            confidences = confidences[indices[:]]
>
> +            classIds = classIds[indices[:]]
>
> +            return bboxes_mlvl, confidences, classIds
>
> +        else:
>
> +            print('nothing detect')
>
>
> Do not print messages here
> ------------------------------
>
> In models/object_detection_nanodet/Nanodet_m_plus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928523287>:
>
> > +            print('nothing detect')
>
> +            return np.array([]), np.array([]), np.array([])
>
> +
>
> +    def distance2bbox(self, points, distance, max_shape=None):
>
> +        x1 = points[:, 0] - distance[:, 0]
>
> +        y1 = points[:, 1] - distance[:, 1]
>
> +        x2 = points[:, 0] + distance[:, 2]
>
> +        y2 = points[:, 1] + distance[:, 3]
>
> +        if max_shape is not None:
>
> +            x1 = np.clip(x1, 0, max_shape[1])
>
> +            y1 = np.clip(y1, 0, max_shape[0])
>
> +            x2 = np.clip(x2, 0, max_shape[1])
>
> +            y2 = np.clip(y2, 0, max_shape[0])
>
> +        return np.stack([x1, y1, x2, y2], axis=-1)
>
> +
>
> +    def drawPred(self, frame, classId, conf, left, top, right, bottom):
>
>
> The drawing function should be placed in demo.py. In this wrapper class,
> we put functions or methods related to inference, pre-processing and
> post-processing.
> ------------------------------
>
> In models/object_detection_nanodet/README.md
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928527188>:
>
> > @@ -0,0 +1,83 @@
>
> +# Nanodet
>
> +
>
> +Nanodet: NanoDet is a FCOS-style one-stage anchor-free object detection model which using Generalized Focal Loss as classification and regression loss.In NanoDet-Plus, we propose a novel label assignment strategy with a simple assign guidance module (AGM) and a dynamic soft label assigner (DSLA) to solve the optimal label assignment problem in lightweight model training.
>
> +
>
> +Note:
>
> +- This version of nanodet: Nanodet-m-plus-1.5x_416
>
> +- The model was trained on COCO 2017 dataset, link to dataset: https://cocodataset.org/#download
>
> +- Below, we have results of COCO data inference
>
> +
>
> +| class         | AP50   | mAP   | class          | AP50   | mAP   |
>
> +|:--------------|:-------|:------|:---------------|:-------|:------|
>
> +| person        | 67.5   | 41.8  | bicycle        | 35.4   | 18.8  |
>
>
> It is good to have AP for different classes as supplements. But I think
> people would care more about the overall mAP. So, please, put the overall
> mAP here and move this table to the end of this file.
> ------------------------------
>
> In models/object_detection_nanodet/demo.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928529937>:
>
> > @@ -0,0 +1,39 @@
>
> +import cv2
>
> +import numpy as np
>
> +import argparse
>
> +import time
>
> +import matplotlib.pyplot as plt
>
> +
>
> +
>
> +from Nanodet_m_plus import nanodet
>
>
> Nanodet_m_plus -> NanodetPlus. NanoDet of different variances should
> share the same pre-processing and post-processing.
> ------------------------------
>
> In models/object_detection_nanodet/demo.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928530507>:
>
> > +try:
>
> +    backends += [cv2.dnn.DNN_BACKEND_TIMVX]
>
> +    targets += [cv2.dnn.DNN_TARGET_NPU]
>
> +    help_msg_backends += "; {:d}: TIMVX"
>
> +    help_msg_targets += "; {:d}: NPU"
>
> +except:
>
> +    print('This version of OpenCV does not support TIM-VX and NPU. Visit https://gist.github.com/Sidd1609/5bb321c8733110ed613ec120c7c02e41 for more information.')
>
> +
>
> +if __name__=='__main__':
>
> +    parser = argparse.ArgumentParser(description='Nanodet inference using OpenCV an contribution by Sri Siddarth Chakaravarthy part of GSOC_2022')
>
> +    parser.add_argument('--image_path', type=str, default='test2.jpg', help="image path")
>
> +    parser.add_argument('--confidence', default=0.35, type=float, help='class confidence')
>
> +    parser.add_argument('--nms', default=0.6, type=float, help='enter nms IOU threshold')
>
> +    args = parser.parse_args()
>
> +
>
> +    image = cv2.imread(args.image_path)
>
>
> Please, add the video stream as input as well
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencv/opencv_zoo/pull/75#pullrequestreview-1048926939>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALTC7V5S3TDGETXR7OTBFBTVVY5UBANCNFSM54QBLGGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Sidd1609 commented on models/object_detection_nanodet/Nanodet_m_plus.py:
```diff
@@ -0,0 +1,158 @@
+        self.project = np.arange(self.reg_max + 1)
+        self.mean = np.array([103.53, 116.28, 123.675], dtype=np.float32).reshape(1, 1, 3)
+        self.std = np.array([57.375, 57.12, 58.395], dtype=np.float32).reshape(1, 1, 3)
+        self.net = cv2.dnn.readNet('nanodet-plus-m-1.5x_416.onnx')
```
Reviewed change, updated filename and replaced changed name in script

Sidd1609 commented on models/object_detection_nanodet/Nanodet_m_plus.py:
```diff
@@ -0,0 +1,158 @@
+import time
+import matplotlib.pyplot as plt
+
+class nanodet():
```
Updated class name to NanoDet

Sidd1609 commented on models/object_detection_nanodet/README.md:
```diff
@@ -0,0 +1,83 @@
+
+| class         | AP50   | mAP   | class          | AP50   | mAP   |
+|:--------------|:-------|:------|:---------------|:-------|:------|
+| person        | 67.5   | 41.8  | bicycle        | 35.4   | 18.8  |
```
Reviewed changes 
Moved the note section to the end, after references.

Sidd1609 commented:
Hey yuantao, will I need to make these changes by Wednesday, as in before
the mentor meetings ?

I have reviewed most of the changes working on the video input now

On Mon, Jul 25, 2022 at 12:50 PM Sri Siddarth Chakaravarthy P <
***@***.***> wrote:

>
> Ok yuantao will update these changes and update the pull request.
>
> On Mon, Jul 25, 2022 at 12:48 PM Yuantao Feng ***@***.***>
> wrote:
>
>> ***@***.**** requested changes on this pull request.
>>
>> Thank you 👍 Please take a look at the comments below.
>>
>> By the way, can you remove the axis of the result images for better
>> visual effects?
>> ------------------------------
>>
>> In models/object_detection_nanodet/Nanodet_m_plus.py
>> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928513752>:
>>
>> > @@ -0,0 +1,158 @@
>>
>> +import cv2
>>
>> +import numpy as np
>>
>> +import argparse
>>
>> +import time
>>
>> +import matplotlib.pyplot as plt
>>
>> +
>>
>> +class nanodet():
>>
>>
>> I think it should be named as NanoDet. So class NanoDet:
>> ------------------------------
>>
>> In models/object_detection_nanodet/Nanodet_m_plus.py
>> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928514977>:
>>
>> > +
>>
>> +class nanodet():
>>
>> +    def __init__(self, prob_threshold=0.35, iou_threshold=0.6):
>>
>> +        with open('coco.names', 'rt') as f:
>>
>> +            self.classes = f.read().rstrip('\n').split('\n')
>>
>> +
>>
>> +        self.num_classes = len(self.classes)
>>
>> +        self.strides = (8, 16, 32, 64)
>>
>> +        self.image_shape = (416, 416)
>>
>> +        self.reg_max = 7
>>
>> +        self.prob_threshold = prob_threshold
>>
>> +        self.iou_threshold = iou_threshold
>>
>> +        self.project = np.arange(self.reg_max + 1)
>>
>> +        self.mean = np.array([103.53, 116.28, 123.675], dtype=np.float32).reshape(1, 1, 3)
>>
>> +        self.std = np.array([57.375, 57.12, 58.395], dtype=np.float32).reshape(1, 1, 3)
>>
>> +        self.net = cv2.dnn.readNet('nanodet-plus-m-1.5x_416.onnx')
>>
>>
>> The model should be renamed to
>> object_detection_nanodet-plus-m-1.5x-416.onnx
>> ------------------------------
>>
>> In models/object_detection_nanodet/Nanodet_m_plus.py
>> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928516320>:
>>
>> > +                img = cv2.copyMakeBorder(img, 0, 0, left, self.image_shape[1] - neww - left, cv2.BORDER_CONSTANT,
>>
>> +                                         value=0)  # add border
>>
>> +            else:
>>
>> +                newh, neww = int(self.image_shape[0] * hw_scale), self.image_shape[1]
>>
>> +                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
>>
>> +                top = int((self.image_shape[0] - newh) * 0.5)
>>
>> +                img = cv2.copyMakeBorder(img, top, self.image_shape[0] - newh - top, 0, 0, cv2.BORDER_CONSTANT, value=0)
>>
>> +        else:
>>
>> +            img = cv2.resize(srcimg, self.image_shape, interpolation=cv2.INTER_AREA)
>>
>> +        return img, newh, neww, top, left
>>
>> +
>>
>> +    def detect(self, srcimg):
>>
>> +        img, newh, neww, top, left = self.resize_image(srcimg)
>>
>> +        img = self._normalize(img)
>>
>> +        blob = cv2.dnn.blobFromImage(img)
>>
>> +        print(blob.shape)
>>
>>
>> Do not print in the inference function.
>> ------------------------------
>>
>> In models/object_detection_nanodet/Nanodet_m_plus.py
>> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928518766>:
>>
>> > +        img = self._normalize(img)
>>
>> +        blob = cv2.dnn.blobFromImage(img)
>>
>>
>> These two lines should be placed in a function called pre_process. Do
>> something like this:
>>
>> def pre_process(self, img):
>>
>>     img = self._normalize(img)
>>
>>     blob = cv2.dnn.blobFromImage(img)
>>
>>     return blob
>>
>> ------------------------------
>>
>> In models/object_detection_nanodet/Nanodet_m_plus.py
>> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928521450>:
>>
>> > +                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
>>
>> +                top = int((self.image_shape[0] - newh) * 0.5)
>>
>> +                img = cv2.copyMakeBorder(img, top, self.image_shape[0] - newh - top, 0, 0, cv2.BORDER_CONSTANT, value=0)
>>
>> +        else:
>>
>> +            img = cv2.resize(srcimg, self.image_shape, interpolation=cv2.INTER_AREA)
>>
>> +        return img, newh, neww, top, left
>>
>> +
>>
>> +    def detect(self, srcimg):
>>
>> +        img, newh, neww, top, left = self.resize_image(srcimg)
>>
>> +        img = self._normalize(img)
>>
>> +        blob = cv2.dnn.blobFromImage(img)
>>
>> +        print(blob.shape)
>>
>> +        self.net.setInput(blob)
>>
>> +
>>
>> +        outs = self.net.forward(self.net.getUnconnectedOutLayersNames())
>>
>> +        det_bboxes, det_conf, det_classid = self.post_process(outs)
>>
>>
>> We need to put pre-process, forward, and post-process in a function
>> called infer for benchmarking. Do something like this:
>>
>> def infer(self, img):
>>
>>     input_blob = self.pre_process(img)
>>
>>
>>
>>     self.net.setInput(input_blob)
>>
>>     outs = self.net.forward(self.net.getUnconnectedOutLayersNames())
>>
>>
>>
>>     bboxes, conf, classid = self.post_process(outs)
>>
>>     return bboxes, conf, classid
>>
>>
>> def detect(self, img):
>>
>>     img, newh, neww, top, left = self.resize_image(img)
>>
>>
>>
>>     det_bboxes, det_conf, det_classid = self.infer(img)
>>
>>
>>
>>     return det_bboxes, det_conf, det_classid
>>
>> self.detect is kept because for now we do not want to have the time of
>> resizing in benchmarking and it is used in demo.py to produce correct
>> results.
>> ------------------------------
>>
>> In models/object_detection_nanodet/Nanodet_m_plus.py
>> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928522492>:
>>
>> > +            bboxes_mlvl /= scale_factor
>>
>> +        scores_mlvl = np.concatenate(scores_mlvl, axis=0)
>>
>> +
>>
>> +        bboxes_wh = bboxes_mlvl.copy()
>>
>> +        bboxes_wh[:, 2:4] = bboxes_wh[:, 2:4] - bboxes_wh[:, 0:2]  ####xywh
>>
>> +        classIds = np.argmax(scores_mlvl, axis=1)
>>
>> +        confidences = np.max(scores_mlvl, axis=1)  ####max_class_confidence
>>
>> +
>>
>> +        indices = cv2.dnn.NMSBoxes(bboxes_wh.tolist(), confidences.tolist(), self.prob_threshold, self.iou_threshold)
>>
>> +        if len(indices)>0:
>>
>> +            bboxes_mlvl = bboxes_mlvl[indices[:]]
>>
>> +            confidences = confidences[indices[:]]
>>
>> +            classIds = classIds[indices[:]]
>>
>> +            return bboxes_mlvl, confidences, classIds
>>
>> +        else:
>>
>> +            print('nothing detect')
>>
>>
>> Do not print messages here
>> ------------------------------
>>
>> In models/object_detection_nanodet/Nanodet_m_plus.py
>> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928523287>:
>>
>> > +            print('nothing detect')
>>
>> +            return np.array([]), np.array([]), np.array([])
>>
>> +
>>
>> +    def distance2bbox(self, points, distance, max_shape=None):
>>
>> +        x1 = points[:, 0] - distance[:, 0]
>>
>> +        y1 = points[:, 1] - distance[:, 1]
>>
>> +        x2 = points[:, 0] + distance[:, 2]
>>
>> +        y2 = points[:, 1] + distance[:, 3]
>>
>> +        if max_shape is not None:
>>
>> +            x1 = np.clip(x1, 0, max_shape[1])
>>
>> +            y1 = np.clip(y1, 0, max_shape[0])
>>
>> +            x2 = np.clip(x2, 0, max_shape[1])
>>
>> +            y2 = np.clip(y2, 0, max_shape[0])
>>
>> +        return np.stack([x1, y1, x2, y2], axis=-1)
>>
>> +
>>
>> +    def drawPred(self, frame, classId, conf, left, top, right, bottom):
>>
>>
>> The drawing function should be placed in demo.py. In this wrapper class,
>> we put functions or methods related to inference, pre-processing and
>> post-processing.
>> ------------------------------
>>
>> In models/object_detection_nanodet/README.md
>> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928527188>:
>>
>> > @@ -0,0 +1,83 @@
>>
>> +# Nanodet
>>
>> +
>>
>> +Nanodet: NanoDet is a FCOS-style one-stage anchor-free object detection model which using Generalized Focal Loss as classification and regression loss.In NanoDet-Plus, we propose a novel label assignment strategy with a simple assign guidance module (AGM) and a dynamic soft label assigner (DSLA) to solve the optimal label assignment problem in lightweight model training.
>>
>> +
>>
>> +Note:
>>
>> +- This version of nanodet: Nanodet-m-plus-1.5x_416
>>
>> +- The model was trained on COCO 2017 dataset, link to dataset: https://cocodataset.org/#download
>>
>> +- Below, we have results of COCO data inference
>>
>> +
>>
>> +| class         | AP50   | mAP   | class          | AP50   | mAP   |
>>
>> +|:--------------|:-------|:------|:---------------|:-------|:------|
>>
>> +| person        | 67.5   | 41.8  | bicycle        | 35.4   | 18.8  |
>>
>>
>> It is good to have AP for different classes as supplements. But I think
>> people would care more about the overall mAP. So, please, put the overall
>> mAP here and move this table to the end of this file.
>> ------------------------------
>>
>> In models/object_detection_nanodet/demo.py
>> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928529937>:
>>
>> > @@ -0,0 +1,39 @@
>>
>> +import cv2
>>
>> +import numpy as np
>>
>> +import argparse
>>
>> +import time
>>
>> +import matplotlib.pyplot as plt
>>
>> +
>>
>> +
>>
>> +from Nanodet_m_plus import nanodet
>>
>>
>> Nanodet_m_plus -> NanodetPlus. NanoDet of different variances should
>> share the same pre-processing and post-processing.
>> ------------------------------
>>
>> In models/object_detection_nanodet/demo.py
>> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r928530507>:
>>
>> > +try:
>>
>> +    backends += [cv2.dnn.DNN_BACKEND_TIMVX]
>>
>> +    targets += [cv2.dnn.DNN_TARGET_NPU]
>>
>> +    help_msg_backends += "; {:d}: TIMVX"
>>
>> +    help_msg_targets += "; {:d}: NPU"
>>
>> +except:
>>
>> +    print('This version of OpenCV does not support TIM-VX and NPU. Visit https://gist.github.com/Sidd1609/5bb321c8733110ed613ec120c7c02e41 for more information.')
>>
>> +
>>
>> +if __name__=='__main__':
>>
>> +    parser = argparse.ArgumentParser(description='Nanodet inference using OpenCV an contribution by Sri Siddarth Chakaravarthy part of GSOC_2022')
>>
>> +    parser.add_argument('--image_path', type=str, default='test2.jpg', help="image path")
>>
>> +    parser.add_argument('--confidence', default=0.35, type=float, help='class confidence')
>>
>> +    parser.add_argument('--nms', default=0.6, type=float, help='enter nms IOU threshold')
>>
>> +    args = parser.parse_args()
>>
>> +
>>
>> +    image = cv2.imread(args.image_path)
>>
>>
>> Please, add the video stream as input as well
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/opencv/opencv_zoo/pull/75#pullrequestreview-1048926939>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ALTC7V5S3TDGETXR7OTBFBTVVY5UBANCNFSM54QBLGGQ>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

fengyuentau commented:
Please, make these changes by Wednesday

Sidd1609 commented:
ok will work on it and update the changes
I have reviewed some of the comments and made the necessary changes, let me
know if you are able to review the changes
- class name changed
- model file name changed
- no print statement in NanodetPlus.py


On Mon, Jul 25, 2022 at 4:07 PM Yuantao Feng ***@***.***>
wrote:

> Please, make these changes by Wednesday
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencv/opencv_zoo/pull/75#issuecomment-1193874155>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALTC7V5RPYCF6BLTGYD26EDVVZU5ZANCNFSM54QBLGGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Sidd1609 commented on models/object_detection_nanodet/Nanodet_m_plus.py:
```diff
@@ -0,0 +1,158 @@
+            classIds = classIds[indices[:]]
+            return bboxes_mlvl, confidences, classIds
+        else:
+            print('nothing detect')
```
reviewed changes 
removed print commands in NanodetPlus.py

Sidd1609 commented on models/object_detection_nanodet/Nanodet_m_plus.py:
```diff
@@ -0,0 +1,158 @@
+            y2 = np.clip(y2, 0, max_shape[0])
+        return np.stack([x1, y1, x2, y2], axis=-1)
+
+    def drawPred(self, frame, classId, conf, left, top, right, bottom):
```
Reviewed changes 
- NanodetPlus.py functions:- softmax_func, infer, pre_process, post_process
- removed drawing function

Sidd1609 commented on models/object_detection_nanodet/demo.py:
```diff
@@ -0,0 +1,39 @@
+import matplotlib.pyplot as plt
+
+
+from Nanodet_m_plus import nanodet
```
reviewed changes 
- changed class name 
- changed file name

Sidd1609 commented on models/object_detection_nanodet/Nanodet_m_plus.py:
```diff
@@ -0,0 +1,158 @@
+        self.net.setInput(blob)
+
+        outs = self.net.forward(self.net.getUnconnectedOutLayersNames())
+        det_bboxes, det_conf, det_classid = self.post_process(outs)
```
reviewed changes
- made changes to the NanodetPlus.py 
- functions: infer, pre_process, and post_process

Sidd1609 commented on models/object_detection_nanodet/Nanodet_m_plus.py:
```diff
@@ -0,0 +1,158 @@
+        img, newh, neww, top, left = self.resize_image(srcimg)
+        img = self._normalize(img)
+        blob = cv2.dnn.blobFromImage(img)
+        print(blob.shape)
```
reviewed changes
- removed print statements in inference function

Sidd1609 commented on models/object_detection_nanodet/Nanodet_m_plus.py:
```diff
@@ -0,0 +1,158 @@
+    def detect(self, srcimg):
+        img, newh, neww, top, left = self.resize_image(srcimg)
+        img = self._normalize(img)
+        blob = cv2.dnn.blobFromImage(img)
```
reviewed changes
- created function "pre-process"
- pre_process function used for image input as well as video input

Sidd1609 commented:
- reviewed all changes 
- no comments left unchecked

Sidd1609 commented on models/object_detection_nanodet/demo.py:
```diff
@@ -0,0 +1,39 @@
+    parser.add_argument('--nms', default=0.6, type=float, help='enter nms IOU threshold')
+    args = parser.parse_args()
+
+    image = cv2.imread(args.image_path)
```
reviewed changes 
- added video inference 
- sample video input data (additional 5 video test data with inference results)
- sample result (additional 15 test images with inference results)

fengyuentau commented on models/object_detection_nanodet/NanodetPlus.py:
```diff
@@ -0,0 +1,139 @@
+        det_bboxes, det_conf, det_classid = self.post_process(outs)
+        ratioh,ratiow = srcimg.shape[0]/newh,srcimg.shape[1]/neww
+
+        return det_bboxes, left, top, ratioh, ratiow, det_bboxes, det_conf, det_classid
```
Double `det_bboxes` in return?

I suggest you put the `letter_box` transformation (the resize and pad processings) in `demo.py`.

Sidd1609 replied:
Reviewed changes
- updated return statement 
- removed additional det_bboxes
- transferred letter_box to main (all bboxes are drawn after inference)

fengyuentau commented on models/object_detection_nanodet/NanodetPlus.py:
```diff
@@ -0,0 +1,139 @@
+                x2 = np.clip(x2, 0, max_shape[1])
+                y2 = np.clip(y2, 0, max_shape[0])
+
+            bboxes = np.stack([x1, y1, x2, y2], axis=-1)
```
Try not to call `np.stack` in a loop. It can be too slow.

Sidd1609 replied:
reviewed changes, 
- replaced np.stack

fengyuentau commented on models/object_detection_nanodet/demo.py:
```diff
@@ -0,0 +1,110 @@
+import numpy as np
+import argparse
+import time
+import matplotlib.pyplot as plt
```
Use `cv2.imshow` to display result image. Take the following as an exmaple:
https://github.com/opencv/opencv_zoo/blob/54190541bb9f717144b255358ca6b37e3e47bf23/models/face_detection_yunet/demo.py#L106-L110

Sidd1609 replied:
reviewed changes, 
- replaced plt.imshow() with cv2.imshow()

fengyuentau commented on models/object_detection_nanodet/demo.py:
```diff
@@ -0,0 +1,110 @@
+    parser = argparse.ArgumentParser(description='Nanodet inference using OpenCV an contribution by Sri Siddarth Chakaravarthy part of GSOC_2022')
+    parser.add_argument('--input_type', type=str, default='image', help="Input types: image or video")
+    parser.add_argument('--image_path', type=str, default='test_data/test1.jpg', help="image path")
+    parser.add_argument('--video_path', type=str, default='test_data/TestV1.mp4', help="video path")
```
My apology for the previous misleading comment. What I meant was taking webcam as input. Take the following as an example:
https://github.com/opencv/opencv_zoo/blob/54190541bb9f717144b255358ca6b37e3e47bf23/models/face_detection_yunet/demo.py#L111-L136

Sidd1609 replied:
reviewed changes, 
- added webcam input

fengyuentau commented on models/object_detection_nanodet/demo.py:
```diff
@@ -0,0 +1,110 @@
+                int((det_bboxes[i,2] - left) * ratiow), srcimg.shape[1]), min(int((det_bboxes[i,3] - top) * ratioh), srcimg.shape[0])
+            frame = drawPred(drawimg, det_classid[i], det_conf[i], xmin, ymin, xmax, ymax)
+
+        plt.imshow(frame)
+        plt.axis('off')
+        plt.show()
```
Do not use plt for result display

fengyuentau commented on models/object_detection_nanodet/demo.py:
```diff
@@ -0,0 +1,110 @@
+        total_frames = 0
+        fps = -1
+
+        model_net = NanoDet(prob_threshold=args.confidence, iou_threshold=args.nms)
```
Remove this repeated line.

Sidd1609 replied:
reviewed changes, 
- removed model initialization

fengyuentau commented on models/object_detection_nanodet/demo.py:
```diff
@@ -0,0 +1,110 @@
+    return frame
+
+if __name__=='__main__':
+    parser = argparse.ArgumentParser(description='Nanodet inference using OpenCV an contribution by Sri Siddarth Chakaravarthy part of GSOC_2022')
```
Make the model path as one of the argument as shown below:
https://github.com/opencv/opencv_zoo/blob/54190541bb9f717144b255358ca6b37e3e47bf23/models/face_detection_yunet/demo.py#L36

Sidd1609 replied:
included modelpath as args "model"

fengyuentau commented on models/object_detection_nanodet/demo.py:
```diff
@@ -0,0 +1,110 @@
+
+            if frame_count >= 50:
+                end = time.time_ns()
+                fps = 1000000000 * frame_count / (end - start)
```
You do not have to calculate FPS manually. Use `cv2.TickMeter()` to measure time, and you can get FPS by calling `getFPS()` as done below:
https://github.com/opencv/opencv_zoo/blob/54190541bb9f717144b255358ca6b37e3e47bf23/models/face_detection_yunet/demo.py#L118-L131

Sidd1609 replied:
reviewed changes, 
- used tm.getFPS()

fengyuentau reviewed (changes requested):
Please, review requested changes below by Wednesday.

You also need to put `test_data` and `results` under a directory named `example`. We do not need videos in examples. It does not make sense to have video examples of size bigger than the model. So, please, remove those videos.

My suggestion is keeping only 2 or 3 images (for image as input), plus 1 gif (for webcam as input) for examples, and display all of them in the `Results` section in README. Each of the image should be like less than 500KB, and the gif should be less than 1MB.

Sidd1609 commented:
Sure will update these changes

On Tue, Jul 26, 2022 at 9:00 AM Yuantao Feng ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> Please, review requested changes below by Wednesday.
>
> You also need to put test_data and results under a directory named example.
> We do not need videos in examples. It does not make sense to have video
> examples of size bigger than the model. So, please, remove those videos.
> And keep only 2 or 3 images (for image as input), plus 1 gif (for webcam as
> input) for examples.
> ------------------------------
>
> In models/object_detection_nanodet/NanodetPlus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929476689>:
>
> > +import argparse
> +import time
> +import matplotlib.pyplot as plt
>
> Remove these lines as they are not used.
> ------------------------------
>
> In models/object_detection_nanodet/NanodetPlus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929477458>:
>
> > +
> +class NanoDet():
> +    def __init__(self, prob_threshold=0.35, iou_threshold=0.6):
> +        with open('coco.names', 'rt') as f:
> +            self.classes = f.read().rstrip('\n').split('\n')
> +
> +        self.num_classes = len(self.classes)
> +        self.strides = (8, 16, 32, 64)
> +        self.image_shape = (416, 416)
> +        self.reg_max = 7
> +        self.prob_threshold = prob_threshold
> +        self.iou_threshold = iou_threshold
> +        self.project = np.arange(self.reg_max + 1)
> +        self.mean = np.array([103.53, 116.28, 123.675], dtype=np.float32).reshape(1, 1, 3)
> +        self.std = np.array([57.375, 57.12, 58.395], dtype=np.float32).reshape(1, 1, 3)
> +        self.net = cv2.dnn.readNet('object_detection_nanodet-plus-m-1.5x-416.onnx')
>
> The model filename should be passed as a parameter (lets name it
> "modelPath" in def __init__.
> ------------------------------
>
> In models/object_detection_nanodet/NanodetPlus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929479496>:
>
> > +        if keep_ratio and srcimg.shape[0] != srcimg.shape[1]:
> +            hw_scale = srcimg.shape[0] / srcimg.shape[1]
> +            if hw_scale > 1:
> +                newh, neww = self.image_shape[0], int(self.image_shape[1] / hw_scale)
> +                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
> +                left = int((self.image_shape[1] - neww) * 0.5)
> +                img = cv2.copyMakeBorder(img, 0, 0, left, self.image_shape[1] - neww - left, cv2.BORDER_CONSTANT,
> +                                         value=0)  # add border
> +            else:
> +                newh, neww = int(self.image_shape[0] * hw_scale), self.image_shape[1]
> +                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
> +                top = int((self.image_shape[0] - newh) * 0.5)
> +                img = cv2.copyMakeBorder(img, top, self.image_shape[0] - newh - top, 0, 0, cv2.BORDER_CONSTANT, value=0)
> +        else:
> +            img = cv2.resize(srcimg, self.image_shape, interpolation=cv2.INTER_AREA)
>
> If the input size is 416x416 exactly, we do not need to resize the source
> image.
> ------------------------------
>
> In models/object_detection_nanodet/NanodetPlus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929481214>:
>
> > +                                         value=0)  # add border
> +            else:
> +                newh, neww = int(self.image_shape[0] * hw_scale), self.image_shape[1]
> +                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
> +                top = int((self.image_shape[0] - newh) * 0.5)
> +                img = cv2.copyMakeBorder(img, top, self.image_shape[0] - newh - top, 0, 0, cv2.BORDER_CONSTANT, value=0)
> +        else:
> +            img = cv2.resize(srcimg, self.image_shape, interpolation=cv2.INTER_AREA)
> +
> +        blob = self.pre_process(img)
> +        self.net.setInput(blob)
> +        outs = self.net.forward(self.net.getUnconnectedOutLayersNames())
> +        det_bboxes, det_conf, det_classid = self.post_process(outs)
> +        ratioh,ratiow = srcimg.shape[0]/newh,srcimg.shape[1]/neww
> +
> +        return det_bboxes, left, top, ratioh, ratiow, det_bboxes, det_conf, det_classid
>
> Double det_bboxes in return?
>
> I suggest you put the letter_box transformation (the resize and pad
> processings) in demo.py.
> ------------------------------
>
> In models/object_detection_nanodet/NanodetPlus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929484178>:
>
> > +
> +            points = anchors
> +            distance = bbox_pred
> +            max_shape=self.image_shape
> +            x1 = points[:, 0] - distance[:, 0]
> +            y1 = points[:, 1] - distance[:, 1]
> +            x2 = points[:, 0] + distance[:, 2]
> +            y2 = points[:, 1] + distance[:, 3]
> +
> +            if max_shape is not None:
> +                x1 = np.clip(x1, 0, max_shape[1])
> +                y1 = np.clip(y1, 0, max_shape[0])
> +                x2 = np.clip(x2, 0, max_shape[1])
> +                y2 = np.clip(y2, 0, max_shape[0])
> +
> +            bboxes = np.stack([x1, y1, x2, y2], axis=-1)
>
> Try not to call np.stack in a loop. It can be too slow.
> ------------------------------
>
> In models/object_detection_nanodet/demo.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929485878>:
>
> > @@ -0,0 +1,110 @@
> +import cv2
> +import numpy as np
> +import argparse
> +import time
> +import matplotlib.pyplot as plt
>
> Use cv2.imshow to display result image. Take the following as an exmaple:
>
> https://github.com/opencv/opencv_zoo/blob/54190541bb9f717144b255358ca6b37e3e47bf23/models/face_detection_yunet/demo.py#L106-L110
> ------------------------------
>
> In models/object_detection_nanodet/demo.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929486900>:
>
> > +def drawPred(frame, classId, conf, left, top, right, bottom):
> +    cv2.rectangle(frame, (left, top), (right, bottom), (0, 0, 0), thickness=2)
> +    #label = '%.2f' % conf
> +    label =''
> +    label = '%s%s' % (classes[classId], label)
> +    labelSize, baseLine = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)
> +    top = max(top, labelSize[1])
> +    # cv.rectangle(frame, (left, top - round(1.5 * labelSize[1])), (left + round(1.5 * labelSize[0]), top + baseLine), (255,255,255), cv.FILLED)
> +    cv2.putText(frame, label, (left, top - 10), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), thickness=2)
> +    return frame
> +
> +if __name__=='__main__':
> +    parser = argparse.ArgumentParser(description='Nanodet inference using OpenCV an contribution by Sri Siddarth Chakaravarthy part of GSOC_2022')
> +    parser.add_argument('--input_type', type=str, default='image', help="Input types: image or video")
> +    parser.add_argument('--image_path', type=str, default='test_data/test1.jpg', help="image path")
> +    parser.add_argument('--video_path', type=str, default='test_data/TestV1.mp4', help="video path")
>
> My apology for the previous misleading comment. What I meant was taking
> webcam as input. Take the following as an example:
>
> https://github.com/opencv/opencv_zoo/blob/54190541bb9f717144b255358ca6b37e3e47bf23/models/face_detection_yunet/demo.py#L111-L136
> ------------------------------
>
> In models/object_detection_nanodet/demo.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929487028>:
>
> > +        plt.imshow(frame)
> +        plt.axis('off')
> +        plt.show()
>
> Do not use plt for result display
> ------------------------------
>
> In models/object_detection_nanodet/demo.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929487543>:
>
> > +
> +        plt.imshow(frame)
> +        plt.axis('off')
> +        plt.show()
> +
> +    else:
> +        model_net = NanoDet(prob_threshold=args.confidence, iou_threshold=args.nms)
> +        colors = [(255, 255, 0), (0, 255, 0), (0, 255, 255), (255, 0, 0)]
> +        capture = cv2.VideoCapture(args.video_path)
> +        start = time.time_ns()
> +
> +        frame_count = 0
> +        total_frames = 0
> +        fps = -1
> +
> +        model_net = NanoDet(prob_threshold=args.confidence, iou_threshold=args.nms)
>
> Remove this repeated line.
> ------------------------------
>
> In models/object_detection_nanodet/demo.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929487854>:
>
> > +with open('coco.names', 'rt') as f:
> +    classes = f.read().rstrip('\n').split('\n')
> +
> +def drawPred(frame, classId, conf, left, top, right, bottom):
> +    cv2.rectangle(frame, (left, top), (right, bottom), (0, 0, 0), thickness=2)
> +    #label = '%.2f' % conf
> +    label =''
> +    label = '%s%s' % (classes[classId], label)
> +    labelSize, baseLine = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)
> +    top = max(top, labelSize[1])
> +    # cv.rectangle(frame, (left, top - round(1.5 * labelSize[1])), (left + round(1.5 * labelSize[0]), top + baseLine), (255,255,255), cv.FILLED)
> +    cv2.putText(frame, label, (left, top - 10), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), thickness=2)
> +    return frame
> +
> +if __name__=='__main__':
> +    parser = argparse.ArgumentParser(description='Nanodet inference using OpenCV an contribution by Sri Siddarth Chakaravarthy part of GSOC_2022')
>
> Make the model path as one of the argument as shown below:
>
> https://github.com/opencv/opencv_zoo/blob/54190541bb9f717144b255358ca6b37e3e47bf23/models/face_detection_yunet/demo.py#L36
> ------------------------------
>
> In models/object_detection_nanodet/demo.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929488933>:
>
> > +
> +            #image = model_net.infer(frame)
> +            srcimg = frame
> +            drawimg = srcimg.copy()
> +            result, left, top, ratioh, ratiow, det_bboxes, det_conf, det_classid = model_net.infer(frame)
> +            for i in range(det_bboxes.shape[0]):
> +                xmin, ymin, xmax, ymax = max(int((det_bboxes[i,0] - left) * ratiow), 0), max(int((det_bboxes[i,1] - top) * ratioh), 0), min(
> +                    int((det_bboxes[i,2] - left) * ratiow), srcimg.shape[1]), min(int((det_bboxes[i,3] - top) * ratioh), srcimg.shape[0])
> +                image = drawPred(drawimg, det_classid[i], det_conf[i], xmin, ymin, xmax, ymax)
> +
> +            frame_count += 1
> +            total_frames += 1
> +
> +            if frame_count >= 50:
> +                end = time.time_ns()
> +                fps = 1000000000 * frame_count / (end - start)
>
> You do not have to calculate FPS manually. Use cv2.TickMeter() to measure
> time, and you can get FPS by calling getFPS() as done below:
>
> https://github.com/opencv/opencv_zoo/blob/54190541bb9f717144b255358ca6b37e3e47bf23/models/face_detection_yunet/demo.py#L118-L131
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencv/opencv_zoo/pull/75#pullrequestreview-1050301686>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALTC7V4FKKI4SCK6TWEA57TVV5LT5ANCNFSM54QBLGGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Sidd1609 commented on models/object_detection_nanodet/NanodetPlus.py:
```diff
@@ -0,0 +1,139 @@
+
+    def infer(self, srcimg, keep_ratio=True):
+        top, left, newh, neww = 0, 0, self.image_shape[0], self.image_shape[1]
+        if keep_ratio and srcimg.shape[0] != srcimg.shape[1]:
+            hw_scale = srcimg.shape[0] / srcimg.shape[1]
+            if hw_scale > 1:
+                newh, neww = self.image_shape[0], int(self.image_shape[1] / hw_scale)
+                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
+                left = int((self.image_shape[1] - neww) * 0.5)
+                img = cv2.copyMakeBorder(img, 0, 0, left, self.image_shape[1] - neww - left, cv2.BORDER_CONSTANT,
+                                         value=0)  # add border
+            else:
+                newh, neww = int(self.image_shape[0] * hw_scale), self.image_shape[1]
+                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
+                top = int((self.image_shape[0] - newh) * 0.5)
+                img = cv2.copyMakeBorder(img, top, self.image_shape[0] - newh - top, 0, 0, cv2.BORDER_CONSTANT, value=0)
+        else:
+            img = cv2.resize(srcimg, self.image_shape, interpolation=cv2.INTER_AREA)
```
The input size is not 416x416 we are passing images without resizing, so I am resizing the images after passing them to inference.

fengyuentau replied:
We need to consider whether the input image is 416x416 already, as we do not want to measure the time of resizing in benchmarking.

Sidd1609 commented:
Ok will check and resize images before running inference

On Tue, Jul 26, 2022 at 3:41 PM Yuantao Feng ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In models/object_detection_nanodet/NanodetPlus.py
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929779830>:
>
> > +        if keep_ratio and srcimg.shape[0] != srcimg.shape[1]:
> +            hw_scale = srcimg.shape[0] / srcimg.shape[1]
> +            if hw_scale > 1:
> +                newh, neww = self.image_shape[0], int(self.image_shape[1] / hw_scale)
> +                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
> +                left = int((self.image_shape[1] - neww) * 0.5)
> +                img = cv2.copyMakeBorder(img, 0, 0, left, self.image_shape[1] - neww - left, cv2.BORDER_CONSTANT,
> +                                         value=0)  # add border
> +            else:
> +                newh, neww = int(self.image_shape[0] * hw_scale), self.image_shape[1]
> +                img = cv2.resize(srcimg, (neww, newh), interpolation=cv2.INTER_AREA)
> +                top = int((self.image_shape[0] - newh) * 0.5)
> +                img = cv2.copyMakeBorder(img, top, self.image_shape[0] - newh - top, 0, 0, cv2.BORDER_CONSTANT, value=0)
> +        else:
> +            img = cv2.resize(srcimg, self.image_shape, interpolation=cv2.INTER_AREA)
>
> We need to consider whether the input image is 416x416 already, as we do
> not want to measure the time of resizing in benchmarking.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencv/opencv_zoo/pull/75#discussion_r929779830>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALTC7V4BQIOGYGIBOGW4ZN3VV62UJANCNFSM54QBLGGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Sidd1609 commented on models/object_detection_nanodet/NanodetPlus.py:
```diff
@@ -0,0 +1,139 @@
+        self.project = np.arange(self.reg_max + 1)
+        self.mean = np.array([103.53, 116.28, 123.675], dtype=np.float32).reshape(1, 1, 3)
+        self.std = np.array([57.375, 57.12, 58.395], dtype=np.float32).reshape(1, 1, 3)
+        self.net = cv2.dnn.readNet('object_detection_nanodet-plus-m-1.5x-416.onnx')
```
Reviewed changes,
- Added args variable for inferring model path (set default path to object_detection_nanodet-plus-m-1.5x-416.onnx)

Sidd1609 commented:
Reviewed changes

fengyuentau commented:
@Sidd1609 I cannot load the onnx model in this pull request. You can reproduce the issue using [netron.app](https://netron.app/) to open your model. Did you manually modify the model?

Sidd1609 commented:
No I did not manually edit the model

On Thu, Jul 28, 2022 at 8:01 AM Yuantao Feng ***@***.***>
wrote:

> @Sidd1609 <https://github.com/Sidd1609> I cannot load the onnx model in
> this pull request. You can reproduce the issue using netron.app to open
> your model. Did you manually modify the model?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencv/opencv_zoo/pull/75#issuecomment-1197581528>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALTC7V2ILPLMEMUL6B7IFNTVWHWIJANCNFSM54QBLGGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Sidd1609 commented:
Shall I share you a google drive link for the model

On Thu, Jul 28, 2022 at 11:34 AM Sri Siddarth Chakaravarthy P <
***@***.***> wrote:

>
> No I did not manually edit the model
>
> On Thu, Jul 28, 2022 at 8:01 AM Yuantao Feng ***@***.***>
> wrote:
>
>> @Sidd1609 <https://github.com/Sidd1609> I cannot load the onnx model in
>> this pull request. You can reproduce the issue using netron.app to open
>> your model. Did you manually modify the model?
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/opencv/opencv_zoo/pull/75#issuecomment-1197581528>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ALTC7V2ILPLMEMUL6B7IFNTVWHWIJANCNFSM54QBLGGQ>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

fengyuentau commented:
> Shall I share you a google drive link for the model

Yes, please send me the model via gdrive and doublecheck on you system as well (e.g. try to read the model with netron or onnx in Python).

Sidd1609 commented:
I am able to read the model on my system using opencv
Cv.dnn.readnet()

On Thu, Jul 28, 2022 at 11:43 AM Yuantao Feng ***@***.***>
wrote:

> Shall I share you a google drive link for the model
>
> Yes, please send me the model via gdrive and doublecheck on you system as
> well (e.g. try to read the model with netron or onnx in Python).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencv/opencv_zoo/pull/75#issuecomment-1197709081>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALTC7V4Z6UPJPEKE627AKRTVWIQJPANCNFSM54QBLGGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Sidd1609 commented:
sending the link for the model
https://drive.google.com/file/d/1FWYWlZm73UiusQpuvxxmiucwMU-YwKVY/view?usp=sharing


On Thu, Jul 28, 2022 at 11:44 AM Sri Siddarth Chakaravarthy P <
***@***.***> wrote:

>
> I am able to read the model on my system using opencv
> Cv.dnn.readnet()
>
> On Thu, Jul 28, 2022 at 11:43 AM Yuantao Feng ***@***.***>
> wrote:
>
>> Shall I share you a google drive link for the model
>>
>> Yes, please send me the model via gdrive and doublecheck on you system as
>> well (e.g. try to read the model with netron or onnx in Python).
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/opencv/opencv_zoo/pull/75#issuecomment-1197709081>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ALTC7V4Z6UPJPEKE627AKRTVWIQJPANCNFSM54QBLGGQ>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

fengyuentau commented:
Tested the model on my side. It works fine on my local system, reaching ~20 FPS with M1 chip. Will do some benchmarking in the following days.

Sidd1609 commented:
Sure, i will begin working on YoloX then.

On Fri, Jul 29, 2022 at 3:26 PM Yuantao Feng ***@***.***>
wrote:

> Tested the model on my side. It works fine on my local system, reaching
> ~20 FPS with M1 chip. Will do some benchmarking in the following days.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencv/opencv_zoo/pull/75#issuecomment-1199090717>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALTC7VYRVO55ZQNBWRROCJDVWOTDFANCNFSM54QBLGGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
